### ISSUE TITLE
Serializing with "$(printf %q "$VARIABLE")"

### ISSUE THREAD

**Issue by stuartpb** [progrium/bashstyle#20]
This is one of those holes that can open up if you're not being conscious of how you're outputting your strings. If it could have spaces, quotes, or dollar signs (and you should _always_ treat strings as potentially having these things), and you're putting it somewhere where it will be treated as part of a string representing an argument list (to be deserialized using `xargs -x`, see #17), you should output it with `printf %q` rather than just naively echoing it (which should also be done with printf, see #18).

**Comment by progrium** [comment]
Cool. This sounds like a good tip that some people might want to wrap in a more explanatory function if they use it a fair amount in their script.

**Comment by progrium** [comment]
Actually, can you give some examples where it's useful to be serializing... including the xargs example, but specifically other cases.

**Comment by stuartpb** [comment]
Saving variables in a generated script, or a list of environment variables like an app's .profile.d/app-env.sh in buildstep/herokuish (which reminds me of another tip/trick for the [High Codex](https://github.com/progrium/bashstyle/issues/16#issuecomment-70016159)).

**Comment by progrium** [comment]
I think I understand what you mean, but some real examples to point me at would be helpful.

**Comment by stuartpb** [comment]
I showed you yesterday how it's used in Pluchu to preserve argument quoting when reproducing the input arguments for ssh's command on the server: https://github.com/plushu/pluchu/blob/master/pluchu#L57

As you could probably guess, Plushu uses it when saving config variables: https://github.com/plushu/plushu-config/blob/master/subcommands/set#L65

Plushu's startup plugin also uses it to export the environment when creating its Upstart unit (but not its systemd unit, I think because systemd doesn't have the same quoting rules as the shell): https://github.com/plushu/plushu-startup-deploy-all/blob/master/install#L24-26

**Comment by progrium** [comment]
Yeah, I get the second two, but I was pretty sure it's unnecessary for the first case as long as it's quoted `"$@"`

**Comment by stuartpb** [comment]
Unfortunately, it is, and this is entirely `ssh`'s fault for sucking at proper argument concatenation. The command `ssh` sends to the server ends up being effectively equivalent to "`$*`", which is why you need to serialize the quoting.

**Comment by progrium** [comment]
So that is a special case for SSH wonkiness.

**Comment by stuartpb** [comment]
Or really any situation where you're going to pass a command onto a shell (such as `docker run ubuntu sh -c`).

**Comment by progrium** [comment]
Alright, I'll have to try this more to internalize it. I believe you're right that it is the right way to universally serialize command line arguments though.

**Comment by progrium** [comment]
So I tried replacing this line:
https://github.com/gliderlabs/herokuish/blob/master/include/procfile.bash#L21

With:
```
unprivileged /bin/bash -c "$(printf '%q' "$@")" 
```

And it was unable to run the commands. Why didn't it work?

**Comment by stuartpb** [comment]
I only took a very brief look at its usage, but isn't the argument to that function already an (incorrectly) serialized argument string?

**Comment by stuartpb** [comment]
Also, I've never had to actually do this within a script (I've only ever passed command strings to xargs to launch something else), but I `think` you can convert a command line string to a list using:

```bash
IFS= read -rd '' -a listname < <(printf %s "$string" | xargs printf '%s\0')
```

**Comment by progrium** [comment]
Yes regarding line 11, though not necessarily (that function can be called directly as a subcommand).

**Comment by stuartpb** [comment]
So yeah, there's your answer, line 21 should really read:

```
unprivileged /bin/bash -c "$1"
```

and then its *callers* should be passing in lists that have been serialized eg. via `printf '%q '` (at some point).

It's like a function that takes JSON. You can pass it `"\"null\""` to pass the string "null", or pass it `"null"` to pass an actual `null`. If somebody passes you `"null"` as your value, you need to understand whether you need to serialize, deserialize, or pass it directly, or you're going to change the value.

**Comment by stuartpb** [comment]
So line 87 in herokuish in bash is the one that should be changed, to this:

```
		/exec)		procfile-exec "$(printf '%q ' "$@")";;
```
